jmx_prometheus_javaagent/src/main/java/io/prometheus/jmx/JavaAgent.java

```diff
@@ -3,54 +3,88 @@
-       port = Integer.parseInt(args[0]);
-       socket = new InetSocketAddress(host, port);
-       file = args[1];
-     }
-
-     new BuildInfoCollector().register();
-     new JmxCollector(new File(file)).register();
-     DefaultExports.initialize();
-     server = new HTTPServer(socket, CollectorRegistry.defaultRegistry, true);
-   }
+    static HTTPServer server;
+
+    public static void agentmain(String agentArgument, Instrumentation instrumentation) throws Exception {
+        premain(agentArgument, instrumentation);
+    }
+
+    public static void premain(String agentArgument, Instrumentation instrumentation) throws Exception {
+        // Bind to all interfaces by default (this includes IPv6).
+        String host = "0.0.0.0";
+
+        try {
+            Config config = parseConfig(agentArgument, host);
+
+            new BuildInfoCollector().register();
+            new JmxCollector(new File(config.file)).register();
+            DefaultExports.initialize();
+            server = new HTTPServer(config.socket, CollectorRegistry.defaultRegistry, true);
+        }
+        catch (IllegalArgumentException e) {
+            System.err.println("Usage: -javaagent:/path/to/JavaAgent.jar=[host:]<port>:<yaml configuration file>");
+            System.exit(1);
+        }
+    }
+
+    /**
+     * Parse the Java Agent configuration. The arguments are typically specified to the JVM as a javaagent as
+     * {@code -javaagent:/path/to/agent.jar=<CONFIG>}. This method parses the {@code <CONFIG>} portion.
+     * @param args provided agent args
+     * @param ifc default bind interface
+     * @return configuration to use for our application
+     */
+    public static Config parseConfig(String args, String ifc) {
+        Matcher matcher = pattern.matcher(args);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException("Malformed arguments - " + args);
+        }
+
+        String givenHost = matcher.group(1);
+        String givenPort = matcher.group(2);
+        String givenConfigFile = matcher.group(3);
+
+        int port = Integer.parseInt(givenPort);
+
+        InetSocketAddress socket;
+        if (givenHost != null && !givenHost.isEmpty()) {
+            socket = new InetSocketAddress(givenHost, port);
+        }
+        else {
+            socket = new InetSocketAddress(ifc, port);
+            givenHost = ifc;
+        }
+
+        return new Config(givenHost, port, givenConfigFile, socket);
+    }
+
+    static class Config {
+        String host;
+        int port;
+        String file;
+        InetSocketAddress socket;
+
+        Config(String host, int port, String file, InetSocketAddress socket) {
+            this.host = host;
+            this.port = port;
+            this.file = file;
+            this.socket = socket;
+        }
+    }
+
+    private static Pattern pattern = Pattern.compile(
+            "^(?:((?:[\\w.]+)|(?:\\[.+])):)" +  // host name, or ipv4, or ipv6 address in brackets
```

Comment:
Escape the `]` too

Reply:
This is considered a redundant escape in Java.